ledger/streaming/client/src/streaming_futures.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Custom Future implementations for the block streaming client
+
+use crate::error::{Error as ClientError, LockReason};
+use futures::{
+    task::{Context, Poll},
+    Future, Stream,
+};
+use mc_ledger_db::Ledger;
+use mc_ledger_streaming_api::{
+    BlockStreamComponents, Error as StreamError, Result as StreamResult,
+};
+use mc_transaction_core::{ring_signature::KeyImage, BlockIndex};
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    sync::{Arc, RwLock},
+};
+
+/// Enum to allow monadic results other than Some & None for
+/// stream types were None would terminate the stream. Useful for
+/// stream combinations like Scan --> Filter_Map
+#[allow(dead_code)]
+pub enum Ready<T> {
+    /// Value with data container to indicate a value is ready
+    Ready(T),
+
+    /// Indicates to subsequent future/stream upstream is not ready
+    NotReady,
+}
+
+/// Future for writing blocks to LedgerDB
+pub struct WriteBlock<L: Ledger> {
+    ledger: Arc<RwLock<L>>,
+    stream_components: Option<BlockStreamComponents>,
+}
+
+impl<L: Ledger> WriteBlock<L> {
+    /// Initialize future with ledger copy & block data
+    pub fn new(ledger: Arc<RwLock<L>>, stream_components: BlockStreamComponents) -> Self {
+        Self {
+            ledger,
+            stream_components: Some(stream_components),
+        }
+    }
+
+    /// future to error Attempt to write block, an error result here results
+    /// in the future returning Poll:Pending
+    fn write_block(&mut self) -> Result<BlockStreamComponents, ClientError> {
+        let block_index: BlockIndex;
+        {
+            let stream_components = self.stream_components.as_ref().unwrap();
+            {
+                let read_only_ledger = self.ledger.try_read()?;
+                block_index = stream_components.block_data.block().index;
+                if read_only_ledger.num_blocks()? + 1 < block_index {
+                    return Err(ClientError::BlockIndexTooFar(block_index));
+                }
+            }
+
+            let mut write_ledger = self.ledger.try_write()?;
+            write_ledger.append_block(
+                stream_components.block_data.block(),
+                stream_components.block_data.contents(),
+                stream_components.block_data.signature().clone(),
+            )?;
+        }
+        Ok(self.stream_components.take().unwrap())
+    }
+}
+
+impl<L: Ledger> Future for WriteBlock<L> {
+    type Output = StreamResult<BlockStreamComponents>;
+
+    ///Attempt to write block and return the block index written
+    fn poll(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.get_mut().write_block() {
```

Comment:
IIUC this will make a blocking I/O call, which is explicitly discouraged as it would stall the async executor from making progress on other tasks.

Reply:
This is something I'm mildly unclear on. In general when you are in an async block, yes, but I believe when you're writing your own futures, you're writing them into a sync context (which then the executor makes into async) I'm not sure that when you write the custom futures futures/tokio::mutex works because you're at that point in the sync context. I haven't looked deeply into the issue so I'll investigate that next.